src/lib/quoter.js

```diff
@@ -1,17 +1,88 @@
 'use strict'
 
 const ILQP = require('ilp').ILQP
+const LiquidityCurve = require('ilp-routing').LiquidityCurve
 const InvalidAmountSpecifiedError = require('../errors/invalid-amount-specified-error')
 const IlpError = require('../errors/ilp-error')
+const log = require('../common/log').create('quoter')
 
 class Quoter {
-  constructor (ledgers) {
+  /**
+   * @param {Ledgers} ledgers
+   * @param {Object} config
+   * @param {Integer} config.quoteExpiry
+   */
+  constructor (ledgers, config) {
     this.ledgers = ledgers
     this.tables = ledgers.tables
+    this.quoteExpiryDuration = config.quoteExpiry // milliseconds
+  }
+
+  /**
+   * @param {Object} request
+   * @param {IlpAddress} request.sourceAccount
+   * @param {IlpAddress} request.destinationAccount
+   * @param {Integer} request.destinationHoldDuration
+   * @param {Boolean} [request._checkCache] default: false
+   * @param {Boolean} [request._shiftCurve] default: true
+   * @returns {Object}
+   */
+  * quoteLiquidity (request) {
+    const liquidityQuote = yield this._quoteLiquidity(
+      Object.assign({_shiftCurve: true}, request))
+    if (!liquidityQuote) return null
+    return Object.assign({}, liquidityQuote, {
+      liquidityCurve: liquidityQuote.liquidityCurve.toBuffer(),
+      expiresAt: new Date(liquidityQuote.expiresAt)
+    })
+  }
+
+  * _quoteLiquidity (request) {
+    const hop = this.tables.localTables.findBestHopForSourceAmount(request.sourceAccount, request.destinationAccount, '0')
+    if (!hop) return Promise.resolve(null)
+    const connector = hop.bestHop
+    const fullRoute = hop.bestRoute
+    const shiftBy = request._shiftCurve
+      ? this.tables.getScaleAdjustment(this.ledgers, fullRoute.sourceLedger, fullRoute.nextLedger)
+      : 0
+    if (isCurveExpired(fullRoute)) {
+      const headRoute = this.tables.localTables.getLocalRoute(fullRoute.sourceLedger, fullRoute.nextLedger)
+      const tailQuote = yield ILQP.quoteByConnector({
+        plugin: this.ledgers.getPlugin(fullRoute.nextLedger),
+        connector,
+        quoteQuery: {
+          destinationAccount: request.destinationAccount,
+          destinationHoldDuration: request.destinationHoldDuration
+        }
+      })
+      if (tailQuote.code) throw new IlpError(tailQuote)
+      const tailCurve = new LiquidityCurve(tailQuote.liquidityCurve)
+      // The quote is more specific than the route.
+      if (fullRoute.targetPrefix.length < tailQuote.appliesToPrefix.length) {
+        log.warn('quote.appliesToPrefix="%s" is more specific than route.targetPrefix="%s"',
+          tailQuote.appliesToPrefix.length, fullRoute.targetPrefix.length)
+      }
+      // The quote is more general than the route, so update the route.
+      fullRoute.curve = headRoute.curve.join(tailCurve)
+      fullRoute.curveExpiresAt = tailQuote.expiresAt.getTime()
+      fullRoute.minMessageWindow = (tailQuote.sourceHoldDuration - request.destinationHoldDuration) / 1000 + headRoute.minMessageWindow
```

Comment:
Right, as discussed on slack, this works because fullRoute is hop.bestRoute passed by reference.
But what if the prefix from the quote result is shorter (more generic) or longer (more specific) than the destinationLedger of fullRoute? Then we want to add a new route into the table, right?